Dirty region related fixes

kakurasan opened the pull request:
* <del>Return D3DERR_INVALIDCALL when an invalid rect is passed to NineTexture9_AddDirtyRect()</del>
* Don't call u_box_union_* when dirty region is empty
    * This fixes broken font rendering issue in WOLF RPG Editor v2.10 games

axeldavy commented on src/gallium/state_trackers/nine/texture9.c:
```diff
@@ -309,6 +309,13 @@ NineTexture9_AddDirtyRect( struct NineTexture9 *This,
         pDirtyRect ? pDirtyRect->left : 0, pDirtyRect ? pDirtyRect->top : 0,
         pDirtyRect ? pDirtyRect->right : 0, pDirtyRect ? pDirtyRect->bottom : 0);
 
+    if (pDirtyRect && (pDirtyRect->left < 0 ||
```
For LockRect, the behaviour is different on XP and Vista+.
I guess it is the same for dirtyRect.
Do you have an example of game that requires the behaviour introduced by this commit ?

axeldavy commented on src/gallium/state_trackers/nine/texture9.c:
```diff
@@ -337,9 +337,13 @@ NineTexture9_AddDirtyRect( struct NineTexture9 *This,
+            rect_to_pipe_box(&box, pDirtyRect);
+            u_box_union_2d(&This->dirty_rect, &This->dirty_rect, &box);
+        }
         if (u_box_clip_2d(&This->dirty_rect, &This->dirty_rect,
```
Interesting fix !

However nine makes other usages of u_box_union_*d which needs to be fixed as well.
Perhaps u_box_union_*d could have a special handling for empty boxes as well... Not sure what is best.
Nine is the only user of the functions, so no problem to modify them.

axeldavy commented on src/gallium/auxiliary/util/u_box.h:
```diff
@@ -124,34 +124,46 @@ static inline void
-   dst->y = y;
+   if (!a->width || !a->height) {
+      memcpy(dst, b, sizeof(struct pipe_box));
+   } else if (!b->width || !b->height) {
```
In the function, you can have dst == a, or dst == b.
memcpy with dst == src seems to be an 'undefined behaviour', which means you have to do another check to avoid the case.
You can also do:
*dst = *b;
I guess it is ok when dst==b ? (not sure 100% sure though)

When the nine state trackers checks for empty box, it justs checks for width == 0. That's because it knows for its use cases checking that is enough (when width is 0, all other fields will be 0 too).

To avoid this issue of aliasing handling, and reduce the number of checks, I tend to think now that the first approach you took was better, ie handle the empty box case in nine. (but applied to all u_box_union_2d usages in nine).

kakurasan committed:
st/nine: Don't call u_box_union_* when dirty region is empty

When dirty region is empty, u_box_union_* incorrectly expands
the new region.

This fixes broken font rendering issue in WOLF RPG Editor v2.10 games.

Signed-off-by: Masanori Kakura <kakurasan@gmail.com>

axeldavy commented on src/gallium/state_trackers/nine/buffer9.c:
```diff
@@ -242,8 +242,13 @@ NineBuffer9_Lock( struct NineBuffer9 *This,
                     nine_csmt_process(This->base.base.device);
-            } else
-                u_box_union_2d(&This->managed.dirty_box, &This->managed.dirty_box, &box);
+            } else {
```
no change needed here, because nine already checks for non-initialized box with managed.dirty.

axeldavy reviewed (changes requested):
Looks good to me (except for the buffer part).

axeldavy commented:
I'll do the change.
